procwrap/procwrap.py

```diff
@@ -2,39 +2,157 @@
 
-def procwrap(command, pid_file=None):
-    """
-    executes the process specified in args.command.
-    returns the returncode of the command.
-    """
-    if pid_file is not None:
-        d = os.path.dirname(pid_file)
-        if not os.path.exists(d):
-            os.makedirs(d)
+# what if this script fails somewhere.
+# say the nsq doesn't connect, etc.
+# we have to use local logging for this.
+import logging
+logging.basicConfig(
+    stream=sys.stderr,
+    datefmt="%Y-%m-%dT%H-%M-%S",
+    format="%(asctime)s %(name)s %(levelname)s %(message)s",
+)
+logger = logging.getLogger()
 
-    process = subprocess.Popen(command)
-    if pid_file is not None:
-        with open(pid_file, "w") as fp:
-            fp.write("%s" % process.pid)
+class ProcessWrapper(object):
+    _NSQD_HTTP_ADDRESS = "http://127.0.0.1:4151"
+    _NSQ_LOG_TOPIC = "log"
+    _DISABLE_NSQ = False
+    _PID_FILE = None
 
-    process.wait()
-    return process.returncode
+    def __init__(
+            self,
+            command,
+            pid_file=_PID_FILE,
+            disable_nsq=_DISABLE_NSQ,
+            nsq_log_topic=_NSQ_LOG_TOPIC,
+            nsqd_http_address=_NSQD_HTTP_ADDRESS):
+
+        self.command = command
+        self.pid_file = pid_file
+        self.disable_nsq = disable_nsq
+        self.nsq_log_topic = nsq_log_topic
+
+        nsqd_http_address = nsqd_http_address.rstrip('/')
+        if not nsqd_http_address.startswith("http://"):
+            nsqd_http_address = 'http://%s' % nsqd_http_address
+
+        self.nsqd_http_address = nsqd_http_address
+
+        self.process = None
+        self.write_to_nsq = lambda msg: None
+
+        if not self.disable_nsq:
+            self.init_nsq()
+
+    def init_nsq(self):
+        session = requests.Session()
+        # TODO use urljoin ?
+        response = None
+        try:
+            response = session.get("%s/ping" % self.nsqd_http_address)
+            if response.status_code != 200:
+                raise Exception("bad response %s" % response)
+        except:
+            logger.exception(
+                "if you wish to disable nsqd, use --disable-nsq"
+            )
+            sys.exit(-1)
+
+        # TODO nsq is a failure point for our logs. what do we do if nsq is down ?
+
+        url = "%s/pub" % self.nsqd_http_address
+        topic = self.nsq_log_topic
+        # TODO what if this write to nsq fails ?
+        self.write_to_nsq = lambda msg: session.post(url, data=msg, params={"topic": topic})
+
+    def start(self):
+        """
+        starts the command and blocks till it finishes.
+        once finished, returns the returncode.
+        """
+
+        if self.pid_file is not None:
+            d = os.path.dirname(self.pid_file)
+            if not os.path.exists(d):
+                os.makedirs(d)
+
+        # NOTE only capturing stderr. stdout still goes to tty.
+
+        self.process = subprocess.Popen(
+            self.command, bufsize=1, universal_newlines=True,
```

Comment:
`bufsize=1` buffers and sends output line by line.
`universal_newlines` is recommended so that all output ending with `\r`, `\r\n`, `\n` is just treated as `\n` so readline works reliably.